src/util/metadata/side_metadata/global.rs

```diff
@@ -157,19 +201,66 @@ impl SideMetadataSpec {
 
-        // yiluowei: Not Sure but this assertion seems too strict for Immix recycled lines
-        #[cfg(not(feature = "global_alloc_bit"))]
-        debug_assert!(start.is_aligned_to(BYTES_IN_PAGE) && meta_byte_lshift(self, start) == 0);
-
         #[cfg(feature = "extreme_assertions")]
         sanity::verify_bzero(self, start, size);
 
-        let meta_start = address_to_meta_address(self, start);
-        if cfg!(target_pointer_width = "64") || self.is_global {
-            memory::zero(
-                meta_start,
-                address_to_meta_address(self, start + size) - meta_start,
+        // Zero for a contiguous side metadata spec. We can simply calculate the data end address, and
+        // calculate the metadata address for the data end.
+        let zero_contiguous = |data_start: Address, data_bytes: usize| {
+            if data_bytes == 0 {
+                return;
+            }
+            let meta_start = address_to_meta_address(self, data_start);
+            let meta_start_shift = meta_byte_lshift(self, data_start);
+            let meta_end = address_to_meta_address(self, data_start + data_bytes);
+            let meta_end_shift = meta_byte_lshift(self, data_start + data_bytes);
+            Self::zero_meta_bits(meta_start, meta_start_shift, meta_end, meta_end_shift);
+        };
+
+        // Zero for a discontiguous side metadata spec (chunked metadata). The side metadata for different
+        // chunks are stored in discontiguous memory. For example, Chunk #2 follows Chunk #1, but the side metadata
+        // for Chunk #2 does not immediately follow the side metadata for Chunk #1. So when we bulk zero metadata for Chunk #1,
+        // we cannot zero up to the metadata address for the Chunk #2 start. Otherwise it may zero unrelated metadata
+        // between the two chunks' metadata.
+        // Instead, we compute how many bytes/bits we need to zero.
+        // The data for which the metadata will be zeroed has to be in the same chunk.
+        #[cfg(target_pointer_width = "32")]
+        let zero_discontiguous = |data_start: Address, data_bytes: usize| {
+            use crate::util::constants::BITS_IN_BYTE;
+            if data_bytes == 0 {
+                return;
+            }
+
+            debug_assert_eq!(
+                data_start.align_down(BYTES_IN_CHUNK),
+                (data_start + data_bytes - 1).align_down(BYTES_IN_CHUNK),
+                "The data to be zeroed in discontiguous specs needs to be in the same chunk"
             );
+
+            let meta_start = address_to_meta_address(self, data_start);
+            let meta_start_shift = meta_byte_lshift(self, data_start);
+
+            // How many bits we need to zero for data_bytes
+            let meta_total_bits = (data_bytes >> self.log_bytes_in_region) << self.log_num_of_bits;
+            let meta_delta_bytes = meta_total_bits >> LOG_BITS_IN_BYTE;
+            let meta_delta_bits: u8 = (meta_total_bits % BITS_IN_BYTE) as u8;
```

Comment:
Nicely done.  I think these lines of code are even clearer than those in `address_to_contiguous_meta_address` and `meta_byte_lshift`.